concordium-smart-contract-testing/src/impls.rs

```diff
@@ -0,0 +1,962 @@
+    ///  - `entrypoint`: the entrypoint to call.
+    ///  - `parameter`: the contract parameter.
+    ///  - `amount`: the amount sent to the contract.
+    ///  - `energy_reserved`: the maximum energy that can be used in the update.
+    pub fn contract_update(
+        &mut self,
+        invoker: AccountAddress,
+        sender: Address,
+        contract_address: ContractAddress,
+        entrypoint: EntrypointName,
+        parameter: OwnedParameter,
+        amount: Amount,
+        energy_reserved: Energy,
+    ) -> Result<SuccessfulContractUpdate, ContractUpdateError> {
+        // Ensure the sender exists.
+        if !self.address_exists(sender) {
+            return Err(ContractUpdateError::SenderDoesNotExist(sender));
+        }
+
+        let invoker_amount_reserved_for_nrg = self.calculate_energy_cost(energy_reserved);
+        // Ensure account exists and can pay for the reserved energy and amount
+        let account_info = self.account_mut(invoker)?;
+        let invoker_signature_count = account_info.signature_count;
+        if account_info.balance.available() < invoker_amount_reserved_for_nrg + amount {
+            return Err(ContractUpdateError::InsufficientFunds);
+        }
+
+        // Ensure that the parameter has a valid size (+2 for the length of parameter).
+        if parameter.0.len() + 2 > constants::MAX_PARAMETER_SIZE {
+            return Err(ContractUpdateError::ParameterTooLarge);
+        }
+
+        // Charge account for the reserved energy up front. This is to ensure that
+        // contract queries for the invoker balance are correct.
+        account_info.balance.total -= invoker_amount_reserved_for_nrg;
+
+        let mut remaining_energy = energy_reserved;
+
+        // Compute the base cost for checking the transaction header.
+        let check_header_cost = {
+            let receive_name = {
+                let contract_name = self
+                    .contracts
+                    .get(&contract_address)
+                    .ok_or(ContractUpdateError::InstanceDoesNotExist(
+                        ContractInstanceMissing(contract_address),
+                    ))?
+                    .contract_name
+                    .as_contract_name();
+                OwnedReceiveName::construct_unchecked(contract_name, entrypoint)
+            };
+
+            let payload = transactions::UpdateContractPayload {
+                amount,
+                address: contract_address,
+                receive_name,
+                message: parameter.clone().into(),
+            };
+
+            let pre_account_trx = transactions::construct::update_contract(
+                invoker_signature_count,
+                invoker,
+                base::Nonce::from(0), // Value does not matter, only used for serialized size.
+                common::types::TransactionTime::from_seconds(0), /* Value does not matter, only
+                                       * used for serialized size. */
```

Comment:
I've added the `size()` helpers to the init/update payload types. But I serialize the data in them to ensure that it matches the format in case we ever decide to use a variable encoding.

Reply:
If we don't expect any changes in the encoding, then I can also just use the known sizes.